### ISSUE TITLE
dapr-operator not starting and TLS issues when running with ArgoCD

### ISSUE THREAD

**Issue by NickLarsenNZ** [dapr/dapr#5066]
## In what area(s)?

> /area runtime

/area operator

/area placement

> /area docs

> /area test-and-release

## What version of Dapr?

1.8.4

## Expected Behavior

After deploying the CRDs and Helm Chart, I expect the operator and ancillary services to stay running.

## Actual Behavior

Sometimes all services come up fine, while other times the pods of the various services are killed due to Readiness Probe failure.

## Steps to Reproduce the Problem

I have not found a way to consistently reproduce this. Last week it happened one time with the sidecar-injector and was simply resolved by restarting it, however this week it is affecting the sidecar-injector, placement-server, and operator.

**values.yml**

```yml
global:
  ha:
    enabled: true
```

**kustomization.yml** (helm install via kustomize and argocd)

```yml
namespace: dapr-system

commonLabels:
  app.kubernetes.io/name: dapr-operator
  app.kubernetes.io/instance: dapr-operator #! Must match the releaseName of the helm-chart, else selectors are broken

resources:
# The apply will fail first time because the CRDs are not there
- https://raw.githubusercontent.com/dapr/dapr/v1.8.4/charts/dapr/crds/components.yaml
- https://raw.githubusercontent.com/dapr/dapr/v1.8.4/charts/dapr/crds/configuration.yaml
- https://raw.githubusercontent.com/dapr/dapr/v1.8.4/charts/dapr/crds/resiliency.yaml
- https://raw.githubusercontent.com/dapr/dapr/v1.8.4/charts/dapr/crds/subscription.yaml

# Unfortunately this needs to stay in the overlay, as they values cannot be overriden if it in the base.
# You can override using kustomize patches (after helm template rendering), but this can complicate
# things when you have to replace large portions such as command/args, and then lose the effect of some
# of the values in helm and have to manually figure out the gaps.
helmCharts:
- name: dapr
  repo: https://dapr.github.io/helm-charts/
  version: v1.8.4
  releaseName: dapr-operator #! see the app.kubernetes.io/instance label
  valuesFile: values.yml
  namespace: dapr-system # can't seem to override the namespace via the ArgoCD Application, and the Helm Chart needs this to populate DNS entries
```

<img width="692" alt="image" src="https://user-images.githubusercontent.com/10092581/186101556-75bf2ee3-9b3e-4f66-99ba-0aed72e8471b.png">

<details>
<summary>working operator logs</summary>

```
time="2022-08-23T08:27:36.322486177Z" level=info msg="log level set to: debug" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T08:27:36.322917161Z" level=info msg="metrics server started on :9090/" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.metrics type=log ver=1.8.4
time="2022-08-23T08:27:36.323613815Z" level=info msg="starting Dapr Operator -- version 1.8.4 -- commit 18575823c74318c811d6cd6f57ffac76d5debe93" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
I0823 08:27:37.374562       1 request.go:665] Waited for 1.0317917s due to client-side throttling, not priority and fairness, request: GET:https://172.20.0.1:443/apis/wgpolicyk8s.io/v1alpha1?timeout=32s
time="2022-08-23T08:27:37.481170494Z" level=info msg="Dapr Operator is starting" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T08:27:37.485102159Z" level=debug msg="observed component to be synced, demo/pubsub" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
I0823 08:27:37.582939       1 leaderelection.go:248] attempting to acquire leader lease dapr-system/operator.dapr.io...
time="2022-08-23T08:27:37.68394938Z" level=info msg="getting tls certificates" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T08:27:37.684072222Z" level=info msg="tls certificates loaded successfully" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T08:27:37.684104051Z" level=info msg="starting gRPC server on port %d6500" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator.api type=log ver=1.8.4
time="2022-08-23T08:27:37.68419769Z" level=info msg="starting watch for certs on filesystem: /var/run/dapr/credentials" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T08:27:37.684615396Z" level=info msg="Dapr Operator started" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T08:27:37.684708418Z" level=info msg="Healthz server is listening on :8080" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T08:27:38.637137082Z" level=info msg="starting webhooks" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
I0823 08:27:38.638127       1 leaderelection.go:248] attempting to acquire leader lease dapr-system/webhooks.dapr.io...
time="2022-08-23T08:27:38.654085142Z" level=info msg="Conversion webhook for \"subscriptions.dapr.io\" is up to date" instance=dapr-operator-6977dccbff-pmr27 scope=dapr.operator type=log ver=1.8.4
```

</details>

<details>
<summary>failed operator logs</summary>

`Readiness probe failed: Get "http://10.220.44.111:8080/healthz": dial tcp 10.220.44.111:8080: connect: connection refused`

> **Note**: log level is set to `info`. I'll replace this with debug logs

```
time="2022-08-23T07:47:47.865564116Z" level=info msg="log level set to: info" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T07:47:47.865702251Z" level=info msg="metrics server started on :9090/" instance=dapr-operator-58d55569db-hd8cb scope=dapr.metrics type=log ver=1.8.4
time="2022-08-23T07:47:47.865813574Z" level=info msg="starting Dapr Operator -- version 1.8.4 -- commit 18575823c74318c811d6cd6f57ffac76d5debe93" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
I0823 07:47:48.916596       1 request.go:665] Waited for 1.039026329s due to client-side throttling, not priority and fairness, request: GET:https://172.20.0.1:443/apis/vpcresources.k8s.aws/v1beta1?timeout=32s
time="2022-08-23T07:47:49.019022457Z" level=info msg="Dapr Operator is starting" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
I0823 07:47:49.119323       1 leaderelection.go:248] attempting to acquire leader lease dapr-system/operator.dapr.io...
time="2022-08-23T07:47:49.220302887Z" level=info msg="getting tls certificates" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T07:47:49.220356142Z" level=info msg="tls certificate not found; waiting for disk changes. err=open /var/run/dapr/credentials/ca.crt: no such file or directory" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T07:47:49.220369295Z" level=info msg="starting watch for certs on filesystem: /var/run/dapr/credentials" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T07:47:50.171631573Z" level=info msg="starting webhooks" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
I0823 07:47:50.172259       1 leaderelection.go:248] attempting to acquire leader lease dapr-system/webhooks.dapr.io...
time="2022-08-23T07:47:50.184331339Z" level=info msg="Conversion webhook for \"subscriptions.dapr.io\" is up to date" instance=dapr-operator-58d55569db-hd8cb scope=dapr.operator type=log ver=1.8.4
time="2022-08-23T07:48:02.732291725Z" level=info msg="Received signal \"terminated\"; beginning shutdown" instance=dapr-operator-58d55569db-hd8cb scope=dapr.signals type=log ver=1.8.4
```

</details>

<details>
<summary>working placement server logs</summary>

> **Note**: log level is set to `info`. I'll replace this with debug logs

```
time="2022-08-23T03:35:15.201648263Z" level=info msg="starting Dapr Placement Service -- version 1.8.4 -- commit 18575823c74318c811d6cd6f57ffac76d5debe93" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T03:35:15.201753237Z" level=info msg="log level set to: info" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T03:35:15.207711991Z" level=info msg="metrics server started on :9090/" instance=dapr-placement-server-1 scope=dapr.metrics type=log ver=1.8.4
time="2022-08-23T03:35:15.261617221Z" level=info msg="Raft server is starting on dapr-placement-server-1.dapr-placement-server.dapr-system.svc.cluster.local:8201..." instance=dapr-placement-server-1 scope=dapr.placement.raft type=log ver=1.8.4
time="2022-08-23T03:35:15.261670151Z" level=info msg="mTLS enabled, getting tls certificates" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T03:35:15.261854035Z" level=info msg="tls certificate not found; waiting for disk changes" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T03:35:15.261957808Z" level=info msg="starting watch for certs on filesystem: /var/run/dapr/credentials" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T03:35:22.292051128Z" level=info msg="tls certificates loaded successfully" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T03:35:22.292176255Z" level=info msg="placement service started on port 50005" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T03:35:22.293548962Z" level=info msg="Healthz server is listening on :8080" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:00:32.664276116Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:00:32.672705915Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:00:32.736339337Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:00:32.736378794Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:00:32.736411383Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:00:34.478336703Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:00:34.486107263Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:14:43.60190176Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:14:43.601962169Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:14:43.602130626Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:23:43.148821472Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:23:43.15314433Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:34:38.308256382Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:34:38.308304248Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:34:38.308342204Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:38:51.773317589Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:38:52.143010602Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:39:03.190696222Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:39:03.190738714Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:39:03.190768212Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:39:04.277996726Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:39:04.283482548Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:15.187828854Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:15.188039428Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:15.188075919Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:34.150229828Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:34.164321214Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:35.39357015Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:35.393625349Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:35.39363869Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:37.011910777Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:51:37.024435131Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:52:25.916553553Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:52:25.916594257Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:52:25.916622743Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:52:42.698829391Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T04:52:42.713754545Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:10:52.362210092Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:10:52.362243559Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:10:52.362269832Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:24:32.180824113Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:24:32.186064118Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:35:42.853253457Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:35:42.853288382Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:35:42.853314947Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:36:26.223682478Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:36:26.228824593Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:41:36.452025091Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:41:36.452079488Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:41:36.452115103Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:41:37.583912915Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:41:37.588575422Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:47:38.375286238Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:47:38.37535024Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T05:47:38.375427481Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T06:05:34.227847412Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T06:05:34.232880506Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T06:16:54.797562541Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T06:16:54.797614157Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T06:16:54.797669137Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T06:52:47.614205109Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T06:52:47.619081353Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:03:42.964180209Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:03:42.964219877Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:03:42.96426149Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:16:07.87873758Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:16:07.884591951Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:27:23.310389181Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:27:23.310439586Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:27:23.310481686Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:33:30.639839451Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:33:30.644880077Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:45:00.762638269Z" level=info msg="shutting down leader loop" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:45:00.762692338Z" level=info msg="Waiting until all connections are drained." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:45:00.762843667Z" level=info msg="cluster leadership lost" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:45:12.699276746Z" level=info msg="cluster leadership acquired" instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T07:45:12.705050352Z" level=info msg="leader is established." instance=dapr-placement-server-1 scope=dapr.placement type=log ver=1.8.4
```

</details>

<details>
<summary>failed placement server logs</summary>

```
time="2022-08-23T08:30:38.555979613Z" level=info msg="starting Dapr Placement Service -- version 1.8.4 -- commit 18575823c74318c811d6cd6f57ffac76d5debe93" instance=dapr-placement-server-2 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T08:30:38.556065151Z" level=info msg="log level set to: debug" instance=dapr-placement-server-2 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T08:30:38.556186223Z" level=info msg="metrics server started on :9090/" instance=dapr-placement-server-2 scope=dapr.metrics type=log ver=1.8.4
time="2022-08-23T08:30:38.563909593Z" level=debug msg="initial configuration%!(EXTRA []interface {}=[index 1 servers [%+v [{Voter dapr-placement-server-0 dapr-placement-server-0.dapr-placement-server.dapr-system.svc.cluster.local:8201} {Voter dapr-placement-server-1 dapr-placement-server-1.dapr-placement-server.dapr-system.svc.cluster.local:8201} {Voter dapr-placement-server-2 dapr-placement-server-2.dapr-placement-server.dapr-system.svc.cluster.local:8201}]]])" instance=dapr-placement-server-2 scope=dapr.placement.raft type=log ver=1.8.4
time="2022-08-23T08:30:38.563964358Z" level=info msg="Raft server is starting on dapr-placement-server-2.dapr-placement-server.dapr-system.svc.cluster.local:8201..." instance=dapr-placement-server-2 scope=dapr.placement.raft type=log ver=1.8.4
time="2022-08-23T08:30:38.563985798Z" level=info msg="mTLS enabled, getting tls certificates" instance=dapr-placement-server-2 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T08:30:38.564087792Z" level=debug msg="entering follower state%!(EXTRA []interface {}=[follower Node at 10.220.42.48:8201 [Follower] leader ])" instance=dapr-placement-server-2 scope=dapr.placement.raft type=log ver=1.8.4
time="2022-08-23T08:30:38.564144351Z" level=info msg="tls certificate not found; waiting for disk changes" instance=dapr-placement-server-2 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T08:30:38.56415761Z" level=info msg="starting watch for certs on filesystem: /var/run/dapr/credentials" instance=dapr-placement-server-2 scope=dapr.placement type=log ver=1.8.4
time="2022-08-23T08:30:39.048100352Z" level=debug msg="accepted connection%!(EXTRA []interface {}=[local-address 10.220.42.48:8201 remote-address 10.220.44.70:49228])" instance=dapr-placement-server-2 scope=dapr.placement.raft type=log ver=1.8.4
time="2022-08-23T08:30:39.04829131Z" level=debug msg="lost leadership because received a requestVote with a newer term%!(EXTRA []interface {}=[])" instance=dapr-placement-server-2 scope=dapr.placement.raft type=log ver=1.8.4
time="2022-08-23T08:30:39.16017251Z" level=debug msg="accepted connection%!(EXTRA []interface {}=[local-address 10.220.42.48:8201 remote-address 10.220.44.70:49232])" instance=dapr-placement-server-2 scope=dapr.placement.raft type=log ver=1.8.4
```

</details>

<details>
<summary>sentry logs (all are up)</summary>

```
time="2022-08-23T08:27:36.035702913Z" level=info msg="starting sentry certificate authority -- version 1.8.4 -- commit 18575823c74318c811d6cd6f57ffac76d5debe93" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:36.035747903Z" level=info msg="log level set to: debug" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:36.037743232Z" level=info msg="metrics server started on :9090/" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.metrics type=log ver=1.8.4
time="2022-08-23T08:27:36.060157002Z" level=info msg="configuration: [port]: 50001, [ca store]: default, [allowed clock skew]: 15m0s, [workload cert ttl]: 24h0m0s" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry.config type=log ver=1.8.4
time="2022-08-23T08:27:36.060189819Z" level=info msg="starting watch on filesystem directory: /var/run/dapr/credentials" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:36.060206817Z" level=info msg="certificate authority loaded" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:36.060671917Z" level=info msg="Healthz server is listening on :8080" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:37.067598229Z" level=info msg="trust root bundle loaded. issuer cert expiry: 2023-08-23 08:27:30 +0000 UTC" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:37.067844436Z" level=info msg="validator created" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:37.067877086Z" level=info msg="sentry certificate authority is running, protecting ya'll" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:27:37.067946401Z" level=debug msg="starting root certificate expiration watcher" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:04.445050572Z" level=debug msg="received issuer credentials changed signal" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:05.445164969Z" level=debug msg="received issuer credentials changed signal" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.445397707Z" level=warning msg="issuer credentials changed; reloading" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.445430523Z" level=info msg="sentry certificate authority is restarting" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.445452968Z" level=info msg="sentry certificate authority is shutting down" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.445597034Z" level=debug msg="terminating root certificate expiration watcher" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.646102805Z" level=info msg="certificate authority loaded" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.661664853Z" level=info msg="root and issuer certs not found: generating self signed CA" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry.ca type=log ver=1.8.4
time="2022-08-23T08:29:06.675734816Z" level=info msg="self signed certs generated and persisted successfully" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry.ca type=log ver=1.8.4
time="2022-08-23T08:29:06.675799829Z" level=info msg="trust root bundle loaded. issuer cert expiry: 2023-08-23 08:29:06 +0000 UTC" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.676069607Z" level=info msg="validator created" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.676107696Z" level=info msg="sentry certificate authority is running, protecting ya'll" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
time="2022-08-23T08:29:06.67617864Z" level=debug msg="starting root certificate expiration watcher" instance=dapr-sentry-574d5c7dd6-hh7wj scope=dapr.sentry type=log ver=1.8.4
```

</details>

## Release Note

RELEASE NOTE:

**Comment by NickLarsenNZ** [comment]
Now that I've enabled debug logging, the operator is coming up. The placement server still fails, so I'll update the logs in the description (with debug level).

**Comment by NickLarsenNZ** [comment]
I'm going to disable liveness probes and see what happens.

**Comment by NickLarsenNZ** [comment]
Nothin appears to be failing anymore, but that doesn't explain why I had seen errors about failing to retrieve TLS certificates, or the K8S API bad certificate.

**Comment by NickLarsenNZ** [comment]
Something else I'm noticing is that when Sentry starts, the Secret `dapr-trust-bundle` goes out of sync (in ArgoCD). When Argo CD Auto-syncs it, it adds in the additional labels:

```yml
apiVersion: v1
data: null
kind: Secret
metadata:
  labels:
    app: dapr-sentry
    app.kubernetes.io/instance: dapr-operator
    app.kubernetes.io/name: dapr-operator
    argocd.argoproj.io/instance: dapr-operator
  name: dapr-trust-bundle
  namespace: dapr-system
```

When Sentry starts up, the secret data changes and the labels are removed:

```yml
apiVersion: v1
data:
  ca.crt: ++++++++
  issuer.crt: ++++++++
  issuer.key: ++++++++
kind: Secret
metadata:
  creationTimestamp: '2022-08-22T09:18:05Z'
  name: dapr-trust-bundle
  namespace: dapr-system
  resourceVersion: '151332687'
  uid: 3bacb2ed-875f-4872-984b-ef51fc2a34f0
type: Opaque

```

I think this might be the problem.

**Comment by NickLarsenNZ** [comment]
Ok, I think I have solved it. It seems at least `dapr-trust-bundle` is being written to and ArgoCD keeps rolling it back to an empty secret.

I have instructed secrets to be ignored from Argo's diff by putting the following into the Application manifest:

```yml
  ignoreDifferences:
  - kind: Secret
    # I might be able to limit it to just that one secret
    # name: dapr-trust-bundle
    jsonPointers:
    - /metadata/labels
    - /data
```

**Comment by NickLarsenNZ** [comment]
I'm going to close this issue for now as it is looking like it is working consistently. I'll re-open if the problem appears again.

**Comment by yaron2** [comment]
@NickLarsenNZ I think we could really use a section in our docs for running Dapr with Argo. Would you like to submit a PR to the docs repo? It will greatly help our community.

**Comment by NickLarsenNZ** [comment]
Hi @yaron2, yes I could give that a go. I'm just having another issue (my original problem with `bad certificate` when the sidecar-injector tries to contact the k8s api. I think it might also be ArgoCD sync related but I can't see where it is happening yet. So, once I am satisfied with my setup, I'd be happy to write some docs.

**Comment by NickLarsenNZ** [comment]
I think I have found where the `http: TLS handshake error from 10.220.48.83:41902: remote error: tls: bad certificate` error (reported by the sidecar-injector` comes from.

I disabled ArgoCD sync, and this morning I woke to find a change it detected (by not auto synced).

```yml
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
  name: dapr-sidecar-injector
webhooks:
  - name: sidecar-injector.dapr.io
    clientConfig:
      caBundle: ... # this value comes up in the diff, yet to be synced
```

When I inspect the certs, I see:

Current CA bundle
```sh
echo -n LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURNakNDQWhxZ0F3SUJBZ0lSQUlWYzZNWmMwSk1LSzM3dU56SFpPTTR3RFFZSktvWklodmNOQVFFTEJRQXcKSXpFaE1COEdBMVVFQXhNWVpHRndjaTF6YVdSbFkyRnlMV2x1YW1WamRHOXlMV05oTUI0WERUSXlNRGd5TkRFdwpNVEkwTWxvWERUTXlNRGd5TVRFd01USTBNbG93SXpFaE1COEdBMVVFQXhNWVpHRndjaTF6YVdSbFkyRnlMV2x1CmFtVmpkRzl5TFdOaE1JSUJJakFOQmdrcWhraUc5dzBCQVFFRkFBT0NBUThBTUlJQkNnS0NBUUVBMUp6bURCVFkKUFY0N0o3ODA0WmlpVjNsdHV4TWNzRHRIOHFwWWFKeEdja2xJUGVqT0JVOXA2ajR2Q0cvZ21DRG13OGczdVkrMwpVNTVmMGc0WnJ1em5vdVQxS0VoY2JDSEo5N2tyLzJLRFQvc3NTbm9JakNVREdXWUxsUEl6dE03R3NCTXRoVmY0Ck1LeGdjQ1BMa2pDek40YlhjVzFobHRzL1hXRzNiUGtUT0l2QjRYTGtHcXpXclExL21wUFo5UGpPakhXcEh4TkEKdGNoUGYydjh0OUM5blo5NlRPenBWUm1wZDd3ZzRPWUtLdVFyNHNsMU8ydzVTcVhZNGpINFVHTk5BdktnMFYyUQpCK2w3aDMyM0RZeDZOcEpjamYvMUZ4VnFYcW50L1AzTytYcloyMVJkRld2T2o5U0ZtRGZIOVNWeHVTaUZkUyt5ClhYVTJVcTBPZlJNcnpRSURBUUFCbzJFd1h6QU9CZ05WSFE4QkFmOEVCQU1DQXFRd0hRWURWUjBsQkJZd0ZBWUkKS3dZQkJRVUhBd0VHQ0NzR0FRVUZCd01DTUE4R0ExVWRFd0VCL3dRRk1BTUJBZjh3SFFZRFZSME9CQllFRkpxYwo5YWZoTDIzcW8xYWU5M2VUV25hRUk1L3RNQTBHQ1NxR1NJYjNEUUVCQ3dVQUE0SUJBUUExbkxORFNQdzJZTkFECkpIMU92ZWRWTjNqbnk0bGxCaWJibitzVkIrbnRJa2pxR2cycm1nU1VHdEFrT2JXTWg3TGoySHM2cmx1M3l2cFUKVEVrZVdnZ1hmZ2dHOFBiMVZ1QUpkZm9rTHovOEpBdTBTLzBVOVY0RXRhMHR1QWJjZUFvNmNtN3QvcUNyZnZYMQpqK2h2a2Nwa1VtMjJ6MTVtVElCZ1BRR3NJRlZTaCtUUVNLT0p6bnVWdTJQMGxaWUdhMzI5NkJDaEdLZFdBaGc1ClV1ZFFkcERHSCszbzhuRnhiWjV4NDZjelFQbzBPQVJGVHZTMlVxNzNkVlppMlNxTC9rNzd0S0FoQW8xallveXQKRXFNT0VDLytxK1NlQ09GRXdIN2lPcGpHbFdJcGZyTUgvM0x6bWVYc3c2NVBjNjZhQU9UUmFaZzNZcnBzb1lGNApHdUswalFDWgotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg== | base64 -d | openssl x509 -noout -subject -dates

subject= /CN=dapr-sidecar-injector-ca
notBefore=Aug 24 10:12:42 2022 GMT
notAfter=Aug 21 10:12:42 2032 GMT
```

... and ArgoCD wants to sync it to:
```sh
echo -n LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURNVENDQWhtZ0F3SUJBZ0lRUkNNUzhMKzJMWjhkVXkxbU04alR3ekFOQmdrcWhraUc5dzBCQVFzRkFEQWoKTVNFd0h3WURWUVFERXhoa1lYQnlMWE5wWkdWallYSXRhVzVxWldOMGIzSXRZMkV3SGhjTk1qSXdPREkxTURNeApOakl5V2hjTk16SXdPREl5TURNeE5qSXlXakFqTVNFd0h3WURWUVFERXhoa1lYQnlMWE5wWkdWallYSXRhVzVxClpXTjBiM0l0WTJFd2dnRWlNQTBHQ1NxR1NJYjNEUUVCQVFVQUE0SUJEd0F3Z2dFS0FvSUJBUUQyMXFLYW1zckYKeHNBWmZ4Y0RQd3NXV0hNKzEzaEQ1WHBydmhpcVdzM05VUUJSMWpaSFZZZVR2OGZKZFdlNXhJUUhkdjF5dmRKWgpyNEVPSzA1c0p6WHVtRjh4eFgxYTlNUDBwRloxT1NMV1I2dTJPTlBIUUVUTGV4b1ZtUlRYWUtueDhsMjlWcDAvCkV4LzlvRDNLd1ZsZ3JIZ01vQURrSjlVUmUxMitwcHkvT0laY2dQNVRKcnBsTFFSYU5INnpkVWpPeDlqeHFtYW0KOWN6NHhrWFRrUTVYOUh4Y2JJWnN1ZWNlZDJmaXZNRzhHV291SjFmMExQM2pBdngydlhhaHhDdVpDeGpoQThrZgp3ZThNU0FydmVMMHpJV0grZ1k3SXZUTEtvU3A5aGFpMy9BQ3laeVRsWkxaVm90Y05HcTEvMkdCbXFjdGVVY0IrCjNCV0NkUFd4U3VRckFnTUJBQUdqWVRCZk1BNEdBMVVkRHdFQi93UUVBd0lDcERBZEJnTlZIU1VFRmpBVUJnZ3IKQmdFRkJRY0RBUVlJS3dZQkJRVUhBd0l3RHdZRFZSMFRBUUgvQkFVd0F3RUIvekFkQmdOVkhRNEVGZ1FVVWZUagpibjNlL1RGVXFETWU4WEtQTysvUzlOY3dEUVlKS29aSWh2Y05BUUVMQlFBRGdnRUJBRmM5TzBXVmRQLy9mRXQrCnVMWE5HRkRQb0VKcjFFdkhrbmFnWjh4SzhaSGFlSE83b2JKWHlpd2cwNlZ0M1pMbk1waUE1amx4amZjNHhITEIKekw4ZU5Jc21JYUowVFFld0xOenRKVDRiWjNVd1JUQ1k4S01RYjE4WHBZSmZORUNLMWd1TXNubDh2VDZtcXdBcwpsay9ibkJ2b05uUzlxUGt1N3dQNGwrWHc5Q0VocGpSd25mS1ZBVFVVNmNHZVBkQkFYb3RjODMxOEJtWStOVktjCjNUM3EyK2I4YlVIVWFEbkVZK1l2MDloSG93OFdXeUpLZitleHdtVHJ1cVIwK0RkWkJIYTRlTVN6RmJVTmY0NUEKWlN1b0oxR2hiV2dOdFZJOWRreWlvckZ6NEpPRWZYZmx2VEc5SzZoVjBaZzhyYytLaDNabmErZkxxdXFyUzNZYwpsMXo3NERnPQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg== | base64 -d | openssl x509 -noout -subject -dates

subject= /CN=dapr-sidecar-injector-ca
notBefore=Aug 25 03:16:22 2022 GMT
notAfter=Aug 22 03:16:22 2032 GMT
```

The dates differ by about 5 hours :thinking:

I'm not sure why it doesn't appear fairly quickly. My assumption is that Argo is checking the manifests in the git repository, and Helm is randomly generating that value. But the out-of-sync alert should be occuring almost immediately.

When I sync the `MutatingWebhookConfiguration`, and kill a pod that has the required `dapr.io` annotations, I immediately see the error reported by the `sidecar-injector`:

```
http: TLS handshake error from 10.220.48.83:41902: remote error: tls: bad certificate
```

I think that is because the webhook is using the previous keypair signed by the previous CA. But the CA trust (`caBundle` for the webhook) has now changed.

I'm just going to get my head around how the webhook and pod update works as the error seems out of place in my mind.

**Comment by NickLarsenNZ** [comment]
This is boiling down to the ca cert and keypair being generated in the Helm chart.

https://github.com/dapr/dapr/blob/20356ee82825c4c051c91ac9e81c2680da9c77a4/charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_webhook_config.yaml#L1-L9

Option 1:
- Advise people to use externally signed certificates so they aren't randomly generated on each `diff` check that ArgoCD does.
- This is arguably more secure anyway since Helm is not generating the certs.
- More effort placed on the people deploying it. It is a bit easier if they have `cert-manager` installed.

Option 2:
- Update the Helm chart so that as the auto generated values change, the sidecar-injector is redeployed. This means moving the ca/keypair generation to a helper. Perhaps the sidecar-injector deployment can then have a label/annotation which is the hash of the `CA` so that it definitely changes.
- There is a moment in time that the injector won't work (the time between the caBundle changing in the `MutatingWebhookConfiguration`, and the cert being used for the webhook in the sidecar-injector).

I think Option 2 should be done anyway, and doesn't limit people who want to go with Option 1.

Any thoughts @yaron2 @ItalyPaleAle?

**Comment by NickLarsenNZ** [comment]
For my own purpose I'm going to implement Option 1 as we can't wait on the release cycle (unless charts changes can go into current releases?). But I'm happy to implement Option 2 when our team is unblocked.

**Comment by NickLarsenNZ** [comment]
On closer inspection, I don't seem to be able to provide my own CA and keypair for the webhook secret and config.

I also now understand this line:
https://github.com/dapr/dapr/blob/20356ee82825c4c051c91ac9e81c2680da9c77a4/charts/dapr/charts/dapr_sidecar_injector/templates/dapr_sidecar_injector_webhook_config.yaml#L38

- Helm will generate the CA and keypair on first install
- On future installs, assuming the MutatingWebhookConfiguration is still in k8s, it will use the same value keypair and caBundle by looking up what is in k8s.

In my case, we're running Helm through Kustomize. Kustomize calls the `helm template` command, which I thought had access to k8s (if configured), but maybe it doesn't. Need to take a closer look at that and see if values are changing on each run.

**Comment by brunocascio** [comment]
I'm facing the same issue, could it be reopened?

Using argocd and it basically rollback the dapr-trust-bundle after first install.

**Comment by NickLarsenNZ** [comment]
@brunocascio , I'll post my workaround for ArgoCD soon. It has been working well for a while now.

**Comment by brunocascio** [comment]
> @brunocascio, in my ArgoCD `Application` manifest for the dapr-operator, I have an `ignoreDifferences` block:
> 
> ```yaml
> apiVersion: argoproj.io/v1alpha1
> kind: Application
> metadata:
>   name: dapr-operator
> spec:
>   destination:
>     name: in-cluster # default
>     namespace: dapr-system
>   source:
>     path: kustomize/.../dapr-operator
>     repoURL: 'https://github.com/example/example'
>     targetRevision: main
>   project: infra
>   syncPolicy:
>     automated:
>       prune: true
>       selfHeal: true
>   ignoreDifferences:
>   - kind: Secret
>     jsonPointers:
>     - /metadata/labels
>     - /data
>   - group: admissionregistration.k8s.io
>     kind: MutatingWebhookConfiguration
>     name: dapr-sidecar-injector
>     jsonPointers:
>     - /webhooks/0/clientConfig/caBundle
> ```
> 
> I'm using Helm inside Kustomize, but the same should work if you're using Helm directly in the ArgoCD app.

Thank you!

I'm using helm with kustomize also, but using applicationset.

I did configure the ignoreDifferences and it's working well so far 👍 What I missed is the `dapr-sidecar-injector ` ignore. I'll add it because I faced the problem of "TLS bad certificate"

Thank you again!

**Comment by igitcode** [comment]
This saved my life today! I configured my secret to just the `dapr-trust-bundle` secret as follows:

```
      ignoreDifferences:
      - kind: Secret
        name: dapr-trust-bundle
        namespace: dapr-system
        jsonPointers:
          - /metadata/labels
          - /data
      - group: admissionregistration.k8s.io
        kind: MutatingWebhookConfiguration
        name: dapr-sidecar-injector
        jsonPointers:
          - /webhooks/0/clientConfig/caBundle
```

**Comment by brunocascio** [comment]
> > @brunocascio, in my ArgoCD `Application` manifest for the dapr-operator, I have an `ignoreDifferences` block:
> > ```yaml
> > apiVersion: argoproj.io/v1alpha1
> > kind: Application
> > metadata:
> >   name: dapr-operator
> > spec:
> >   destination:
> >     name: in-cluster # default
> >     namespace: dapr-system
> >   source:
> >     path: kustomize/.../dapr-operator
> >     repoURL: 'https://github.com/example/example'
> >     targetRevision: main
> >   project: infra
> >   syncPolicy:
> >     automated:
> >       prune: true
> >       selfHeal: true
> >   ignoreDifferences:
> >   - kind: Secret
> >     jsonPointers:
> >     - /metadata/labels
> >     - /data
> >   - group: admissionregistration.k8s.io
> >     kind: MutatingWebhookConfiguration
> >     name: dapr-sidecar-injector
> >     jsonPointers:
> >     - /webhooks/0/clientConfig/caBundle
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm using Helm inside Kustomize, but the same should work if you're using Helm directly in the ArgoCD app.
> 
> Thank you!
> 
> I'm using helm with kustomize also, but using applicationset.
> 
> I did configure the ignoreDifferences and it's working well so far 👍 What I missed is the `dapr-sidecar-injector ` ignore. I'll add it because I faced the problem of "TLS bad certificate"
> 
> Thank you again!

Well, it happened again after 25 days running

**Comment by NickLarsenNZ** [comment]
> Well, it happened again after 25 days running

@brunocascio What changed? Was something missing from the paths in the above ignoreDifferences configuration?

**Comment by brunocascio** [comment]
> > Well, it happened again after 25 days running
> 
> @brunocascio What changed? Was something missing from the paths in the above ignoreDifferences configuration?

Nothing changed, the same configurations I mentioned above.
After a while, it stopped working again with the same "bad tls certificate" error.

**Comment by brunocascio** [comment]
@NickLarsenNZ I'm still facINGIN this issue even after switching Dapr as a separate application, so the setup is the same as yours.

I'm creating a `Component` inside of dapr-system, so everytime we upgrade that component the error with certificates appears. 

I did set self-signed certificates with 10 years validity (dapr-trust-bundle secret) but the error still happen. 

Could we reopen this issue maybe?

**Comment by brunocascio** [comment]
For reference, I did make some changes in my Argo application, so now is working properly. Here's my full config:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: dapr
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  generators:
    - list:
        elements:
          - cluster: nonprod-workload01
  template:
    metadata:
      name: 'dapr-{{cluster}}'
    spec:
      project: dapr
      destination:
        name: '{{cluster}}'
        namespace: dapr-system
      sources:
        - repoURL:  https://repo.com/kube/cluster-apps.git
          path: 'bootstrap/dapr/{{cluster}}'
          targetRevision: HEAD
      syncPolicy:
        automated:
          prune: true
          selfHeal: true
        syncOptions:
          - RespectIgnoreDifferences=true      # <---- added this
      ignoreDifferences:
        - kind: Secret
          jsonPointers:
          - /metadata/labels
          - /data
        - group: admissionregistration.k8s.io
          kind: MutatingWebhookConfiguration
          name: dapr-sidecar-injector
          jsonPointers:
          - /webhooks/0/clientConfig/caBundle
```

Basically, `RespectIgnoreDifferences=true ` fixed the problem

Sometimes we use "Hard Refresh" on the application, which means Argo will fetch and make the diff against all the manifests where those are ignored, but at sync stage, it's applied as it is so that was the problem I was facing. Setting up the flag `RespectIgnoreDifferences` the problem was solved because argo also ignore the changes on sync stage as well.

That's explained here https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#respect-ignore-difference-configs.

**Comment by Pionerd** [comment]
So I have seen other helm chart (e.g. Nginx-ingress, https://github.com/kubernetes/ingress-nginx/blob/main/charts/ingress-nginx/templates/admission-webhooks/cert-manager.yaml) where the certificate gets generated by Cert Manager instead of relying on helm to generate the self signed certificate. This no longer leads to argoCD certs out of sync. Has anyone considered building this into the daps helm chart?
